### PR TITLE
Implemented selection type non white space sequence

### LINKED ISSUE #330: Add `nonWhitespaceSequence` scope type
We can call it "paint", and it will expand to include all non-whitespace characters adjacent to selection.  Can proceed by:

1. Get line containing end of selection
2. Move end forward until just before the first whitespace character you encounter
3. Get line containing start of selection
4. Move start back until just after the first whitespace character you encounter

Question: what to do if one of the endpoints is in the middle of whitespace?  Which of the following do we do:

- Leave the endpoint untouched
- Move it forward until the end of the next non-whitespace block
- Move it backward to the end of the preceding non-whitespace block

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
fixes #330

### CODE REVIEW

File: src/processTargets/processSelectionType.ts [hunk 1]
```diff
@@ -173,6 +175,78 @@ function processParagraph(
...
+  const { selectionType, insideOutsideType, position, modifier } = target;
+  const { document } = selection.editor;
+  const { start, end } = selection.selection;
+  let newSelection;
+
+  let startLine;
+  let i = start.line;
+  do {
+    startLine = document.lineAt(i++);
+  } while (startLine.isEmptyOrWhitespace && i <= end.line);
```
Comment (pokey) [inline src/processTargets/processSelectionType.ts:192]: Why can't you just use `start.line` directly?

File: src/test/suite/fixtures/recorded/selectionTypes/takePaint.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+languageId: plaintext
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/selectionTypes/takePaint.yml:1]: Based on our discussion, this one should be an error, no?

File: src/test/suite/fixtures/recorded/selectionTypes/clearPaint.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+languageId: plaintext
```
Comment (pokey) [inline src/test/suite/fixtures/recorded/selectionTypes/clearPaint.yml:1]: I don't understand this test. Why is this not an error?

File: src/processTargets/processSelectionType.ts [hunk 4]
```diff
@@ -173,6 +175,78 @@ function processParagraph(
...
+
+  // Can't find any non whitespace characters in the selection
+  if (startLine.isEmptyOrWhitespace) {
```
Comment (pokey) [inline src/processTargets/processSelectionType.ts:195]: This implementation seems overly complicated. What's wrong with the approach proposed in the issue description?

**Review by pokey — COMMENTED** [review]
Thanks for taking the initiative on this one, but not sure bout the impl. I was expecting it to end up a lot simpler. Is there some corner case I didn't think of?

File: src/typings/Types.ts [hunk 5]
```diff
@@ -186,7 +186,12 @@ export type Modifier =
 
 export type SelectionType =
   //   | "character" Not implemented
-  "token" | "line" | "notebookCell" | "paragraph" | "document";
+  | "token"
+  | "line"
+  | "notebookCell"
+  | "paragraph"
+  | "document"
+  | "nonWhitespaceSequence";
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply src/typings/Types.ts:194]: Exactly according to the implementation we have today selection types are more based around text syntax

File: src/processTargets/processSelectionType.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: Because that one might be empty. This solves the corner case where one of the edges are in an empty part of the document

File: src/test/suite/fixtures/recorded/selectionTypes/clearPaint.yml (hunk 3 above)
Comment (AndreasArvidsson) [reply src/test/suite/fixtures/recorded/selectionTypes/clearPaint.yml:1]: Why would it? Are referring to my issue where all plain text tests fail?

File: src/processTargets/processSelectionType.ts (hunk 4 above)
Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:195]: I would say it follows the issue exactly. . The implementation needs to be this complex to get that functionality

File: src/processTargets/processSelectionType.ts (hunk 1 above)
Comment (pokey) [reply src/processTargets/processSelectionType.ts:192]: If it's empty, then the start is not part of a non-white space sequence so that should be an error

File: src/typings/Types.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (pokey) [reply src/typings/Types.ts:194]: Ok that's prob fine for now then. That distinction is gonna be removed soon anyways

File: src/test/suite/fixtures/recorded/selectionTypes/clearPaint.yml (hunk 3 above)
Comment (pokey) [reply src/test/suite/fixtures/recorded/selectionTypes/clearPaint.yml:1]: See above; the end of selection is not contained in a non-white space sequence, so should be an error

File: src/processTargets/processSelectionType.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: I've implemented this port
```
Move it forward until the end of the next non-whitespace block
Move it backward to the end of the preceding non-whitespace block
```
I think that makes total sense and the implementation as it is right now is really useful in my opinion

Comment (pokey) [reply src/processTargets/processSelectionType.ts:192]: Yes but we discussed yesterday and decided to throw an error if an endpoint is not touching a non-whitespace sequence

Comment (pokey) [reply src/processTargets/processSelectionType.ts:192]: Also I don't think "take paint file" will work today, will it? Can't specify two selection types

Comment (pokey) [reply src/processTargets/processSelectionType.ts:192]: And "take paint file" violates the semantics of saying a scope type, which implies smallest containing instance of that scope type, not largest contained instance

Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: That's misunderstanding I said that we should throw an error if it was in complete white space. that the ends are is fine I think.

It will work with new grammar. It's already implemented and is useful for the future.

Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: We already have sub word modifier. I see no reason why modify can't both grow and shrink. 
A is now paint selects all content the current selection touches.

Comment (pokey) [reply src/processTargets/processSelectionType.ts:192]: So would you argue that if your selection is larger than a function, it should shrink to the function? What happens if it is larger than one function but contained in another? Just seems a bit ambiguous to allow "funk" to both shrink and expand. To me the semantics of `containingScope` are that it expands

Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: But I haven't implemented this as a containing scope. It is a selection type.

Containing scopes by their name is a bit different. But definitely to select all functions that your selection touches would probably be useful.

Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: But I'm not married to the solution so if you hate it in can change you to something simpler

Comment (pokey) [reply src/processTargets/processSelectionType.ts:192]: In the future there will be no distinction between containing scope and selection type so I'd just treat this as if it were selection type

I guess I'm not opposed to changing the semantics of containing scope to just be every instance of that scope that is touched by the input, just want to make sure we're consistent

I think that's actually pretty close to what we do with "token", right?

Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: I know but in this future we don't have to stick to the name containing scope if we don't want to. 

Yes we token we are growing but we don't shrink. But the question is if we shouldn't treat all the scope modifiers like this instead? I think we have to have a proper discussion about it

Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: To give some context the reason I think this fits particularly well with paint is that we already allow start and end to be in different sequences. Something we don't do containing scopes. If we already decided that this one is going to be treated differently I think allowing it to shrink is totally fine even if we don't implement it on the others scopes. I think we ide can have it as wildcard with its own rules or it should adhere to the same as containing scopes where both ends of the selection has to be inside and an error is thrown if it's not.

Comment (pokey) [reply src/processTargets/processSelectionType.ts:192]: We could def discuss, but fwiw I'm happy to just agree to change the semantics of today's `containingScope` modifier to be "any instance of the scope type touching the input to the modifier", which would necessitate your impl here, and then we could just merge.  That way we don't need to block on having a discussion.  But if you wanna talk it through before we merge I'm of course happy to do that

The one catch is that we'll then want to change impl of "round", "funk", etc to be consistent with these semantics.  Probably not urgent because it's kind of a corner case but just a thought

We could just merge this one, then file an issue to make sure to discuss / think through the semantics of containing scope (or whatever we rename it to to reflect that it's really "touching scope")

Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts:192]: I want to have a quick discussion about it before. I'm available before today's meeting if you are?

Comment (pokey) [reply src/processTargets/processSelectionType.ts:192]: As discussed, let's expand start cursor to containing sequence and end cursor to containing sequence, but never shrink, just throwing an error of either end is not contained in a non-whitespace sequence

Comment (AndreasArvidsson) [reply src/processTargets/processSelectionType.ts]: done
